be/src/formats/parquet/file_writer.cpp

```diff
@@ -194,140 +313,635 @@ Status FileWriterBase::init() {
+                                             const ::parquet::schema::NodePtr& node, const ColumnPtr& col) {
+    DCHECK(type_desc.type == TYPE_STRUCT);
+    auto struct_node = std::static_pointer_cast<::parquet::schema::GroupNode>(node);
+
+    unsigned char* nulls = nullptr;
+    if (col->is_nullable()) {
+        const auto null_column = down_cast<NullableColumn*>(col.get())->null_column();
+        nulls = null_column->get_data().data();
     }
 
+    const auto data_column = ColumnHelper::get_data_column(col.get());
+    const auto struct_column = down_cast<StructColumn*>(data_column);
+    Context derived_ctx(ctx._max_def_level + node->is_optional(), ctx._max_rep_level, ctx.size() + data_column->size());
+
+    int subcol_size = 0;
+    for (auto i = 0; i < ctx.size(); i++) {
+        auto [idx, def_level, rep_level] = ctx.get(i);
+        if (idx == Context::kNULL) {
+            derived_ctx.append(Context::kNULL, def_level, rep_level);
+            continue;
+        }
+        if (nulls != nullptr && nulls[idx]) {
+            derived_ctx.append(Context::kNULL, def_level, rep_level);
+            subcol_size++;
+            continue;
+        }
+        derived_ctx.append(subcol_size++, def_level + node->is_optional(), rep_level);
+    }
+
+    for (size_t i = 0; i < type_desc.children.size(); i++) {
+        auto sub_col = struct_column->field_column(type_desc.field_names[i]);
+        DCHECK(sub_col->size() == subcol_size);
+        auto ret = _add_column_chunk(derived_ctx, type_desc.children[i], struct_node->field(i), sub_col);
+        if (!ret.ok()) {
+            return ret;
+        }
+    }
     return Status::OK();
 }
 
-// The current row group written bytes = total_bytes_written + total_compressed_bytes + estimated_bytes.
-// total_bytes_written: total bytes written by the page writer
-// total_compressed_types: total bytes still compressed but not written
-// estimated_bytes: estimated size of all column chunk uncompressed values that are not written to a page yet. it
-// mainly includes value buffer size and repetition buffer size and definition buffer value for each column.
-std::size_t FileWriterBase::_get_current_rg_written_bytes() const {
-    if (_rg_writer == nullptr) {
-        return 0;
+std::vector<uint8_t> ChunkWriter::_make_null_bitset(size_t n, const uint8_t* nulls) const {
+    // TODO(letian-jiang): optimize
+    DCHECK(nulls != nullptr);
+    std::vector<uint8_t> bitset((n + 7) / 8);
+    for (size_t i = 0; i < n; i++) {
+        if (!nulls[i]) {
+            bitset[i / 8] |= 1 << (i % 8);
+        }
     }
+    return bitset;
+}
 
-    auto estimated_bytes = std::accumulate(_buffered_values_estimate.begin(), _buffered_values_estimate.end(), 0);
-
-    return _rg_writer->total_bytes_written() + _rg_writer->total_compressed_bytes() + estimated_bytes;
+std::vector<int16_t> ChunkWriter::_make_def_levels(const Context& ctx, const uint8_t* nulls) const {
```

Comment:
you could pass vector<int16_t>& to this method in order to reduce memory copy

Reply:
done